File: android/app/src/main/java/com/dam/starwarsapp/data/remote/dto/FilmDto.kt
package com.dam.starwarsapp.data.remote.dto

import com.squareup.moshi.Json

data class FilmDto(
    val title: String,
    @Json(name = "episode_id") val episodeId: Int,
    @Json(name = "opening_crawl") val openingCrawl: String,
    val director: String,
    val producer: String,
    @Json(name = "release_date") val releaseDate: String,
    val characters: List<String> = emptyList(),
    val planets: List<String> = emptyList(),
    val starships: List<String> = emptyList(),
    val vehicles: List<String> = emptyList(),
    val species: List<String> = emptyList(),
    val created: String,
    val edited: String,
    val url: String,
)

File: iOS/Star-Wars-Demo-APP/Star-Wars-Demo-APP/Data/Remote/DTO/FilmDTO.swift
import Foundation

struct FilmDTO: Decodable {
    let title: String
    let episodeId: Int
    let openingCrawl: String
    let director: String
    let producer: String
    let releaseDate: String
    let characters: [String]
    let planets: [String]
    let starships: [String]
    let vehicles: [String]
    let species: [String]
    let url: String

    enum CodingKeys: String, CodingKey {
        case title, director, producer, characters, planets, starships, vehicles, species, url
        case episodeId = "episode_id"
        case openingCrawl = "opening_crawl"
        case releaseDate = "release_date"
    }
}

extension FilmDTO {
    func toDomain() -> Film {
        let id = url.split(separator: "/").last.flatMap { Int($0) } ?? 0
        return Film(
            id: id,
            title: title,
            episodeId: episodeId,
            openingCrawl: openingCrawl,
            director: director,
            producer: producer,
            releaseDate: releaseDate,
            charactersCount: characters.count,
            planetsCount: planets.count,
            starshipsCount: starships.count,
            vehiclesCount: vehicles.count,
            speciesCount: species.count
        )
    }
}
